Adjust osu!taiko and osu!mania editors to not visualise velocity changes by default

smoogipoo opened the pull request:
Stable never had this, and it's more natural to edit in linear speed.

I thought it would be interesting to have a menu item toggle for this but it appears like we can no longer change the visualisation method after a ruleset's been loaded.

https://github.com/ppy/osu/assets/191335/07512a9b-fc6c-4255-b0e4-5508bf5e9e4a

smoogipoo committed:
Disable SVs from being visualised in mania editor

peppy commented:
On the fence about this.... one of the goals of the editor was to be a layer on top of gameplay, and this strictly goes against that concept. Can understand it making editing easier, but like, previewing is no longer a seamless thing, and you have no idea how a beatmap is going to look / play without changing display modes via a new screen.

I feel like this would need to be a very visible toggle, alongside a zoom control.

Have you investigated how hard it would be to change the algorithm on the fly?

smoogipoo committed:
Allow toggling SVs in the editor

smoogipoo commented:
From a technical perspective it isn't difficult to have it toggleable: https://github.com/smoogipoo/osu/commit/9ce84e6ee144931627ec9a73a0bc9c5db2910b0f

But I'm not sure about the UX in the above branch.

I'd really like to not spend too much time on this, because the editor is pretty far down on my priority list - this change is 1 line of code that constitutes what I think is an MVP. Mappers have used constant scroll with the same UX in the osu!stable editor for about a decade by now.

peppy commented:
I guess the next question is: should this also be applied to taiko? catch? Should it even be applied to osu!mania when there hasn't been a single issue/discussion complaining yet?

Would like to see a bit more discussion and have a feel of it myself before making any abrupt decisions.

smoogipoo commented:
> Should it even be applied to osu!mania when there hasn't been a single issue/discussion complaining yet?

Consider this the discussion thread, and that I've done this only because it's been brought up to me.

> should this also be applied to taiko?

Actually, it _must_ be applied to taiko, as it's impossible to map with SVs there due to their overlapping algorithm. I thought it was already applied but it appears not...

peppy commented:
I'd definitely be more okay if it's applied everywhere, put it that way.

DannyPX commented:
A toggle for SV visualization would be nice to preview the map. But it would be practically too difficult to edit the map with SV visualizations, especially on heavier SV maps like maps in tournaments.

As for a feature in the future, having a preview playfield next to the editor playfield, and a toggle to turn on/off the preview. This is done in certain rhythm games already, like Quaver. 
![image](https://github.com/ppy/osu/assets/47347980/fdbcfb85-cc4b-4fcb-851c-40fc06a87681)

peppy commented:
> having a preview playfield next to the editor playfield, and a toggle to turn on/off the preview.

The timeline already acts as a preview of the constant timing layout. The main area is a preview of the gameplay. That was my intended direction forward.

smoogipoo commented:
> On a brief check, it seems like the taiko editor handles quite nicely even with the algorithm set to overlapping

It depends on the map and if there are timing points with significant enough speed to overlap. For example (and this is not an extreme one): https://dev.ppy.sh/beatmapsets/1603159#taiko/3275829

https://github.com/ppy/osu/assets/1329837/d57b35b3-1307-44ea-a1b7-6f55ad22ba35

peppy committed:
Merge branch 'master' into mania-edit-disable-sv

peppy committed:
Expose base `VisualisationMethod` so we don't need to `new` it locally per ruleset

peppy committed:
Move toggle implementation to work on all scrolling rulesets automatically

peppy commented:
I've pushed changed to make this work across all rulesets, but taiko needs consideration. I may revert this change and keep it local to osu!mania if I can't figure a solution quickly.

peppy committed:
Merge branch 'master' into mania-edit-disable-sv

peppy committed:
Change implementation of "show speed changes" to require explicit ruleset support

peppy committed:
Remvoe null default for mods which can't be null

peppy committed:
Implement new interface-based speed change visualisation support on mania/taiko

peppy commented:
@smoogipoo I came up with a good in-between solution. Can you check and make sure it looks okay to your eye?

bdach committed:
Move comment a bit to fix formatting

Would otherwise trigger IDE0055, but that isn't resolveable without
an inspection cycle with resharper, so just move in a more sane place.

bdach commented on osu.Game/Rulesets/UI/Scrolling/DrawableScrollingRuleset.cs:
```diff
@@ -178,6 +163,36 @@ protected override void LoadComplete()
+            }
+        }
+
+        private void updateScrollAlgorithm()
```
Out of scope of this pull, but I majorly like this, since it will help fix issues like SVs not taking effect immediately pretty much instantly (just subscribe to `ControlPoints` updates and call this).

bdach commented on osu.Game/Rulesets/Edit/ScrollingHitObjectComposer.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace osu.Game.Rulesets.Edit
+{
+    public abstract partial class ScrollingHitObjectComposer<TObject> : HitObjectComposer<TObject>
```
This though I'm feeling kind of bearish about since it's starting to feel like a potential future inheritance finger trap. But I trust that we'll be able to recognise when it gets bad enough so probably fine for now...

peppy replied:
yeah.

bdach reviewed (approved):
Not @smoogipoo but figure I'll drop a review so that this can maybe get into release I guess

smoogipoo commented:
The changes look good to me.

bdach committed:
Merge branch 'master' into mania-edit-disable-sv
